addons/website_sale/controllers/main.py

```diff
@@ -194,29 +186,24 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
 
         keep = QueryURL('/shop', category=category and int(category), search=search, attrib=attrib_list)
 
-        if not context.get('pricelist'):
+        if not local_context.get('pricelist'):
             pricelist = self.get_pricelist()
-            context['pricelist'] = int(pricelist)
+            local_context['pricelist'] = pricelist.id
         else:
-            pricelist = pool.get('product.pricelist').browse(cr, uid, context['pricelist'], context)
+            pricelist = request.env['product.pricelist'].browse(local_context['pricelist'])
         url = "/shop"
         if search:
             post["search"] = search
         if category:
-            category = pool['product.public.category'].browse(cr, uid, int(category), context=context)
+            category = category.with_context(local_context)
```

Comment:
we make a `browse(int(category))` because it is possible that category is not a recordset but a string (`/shop?category=42` => `category='42'`), not to browse it with pricelist context, see d978c7824614a22048c140ffaa6b0256c6405542